File: src/funcs/apis/ytmusicsearchapi.py
import argparse
import json
from ytmusicapi import YTMusic

# Initialize the YTMusic API
ytmusic = YTMusic()

def get_track_list(content_id, content_type):
    """
    Get track list for an album or playlist.

    Parameters:
        content_id (str): The ID of the album or playlist
        content_type (str): Either 'album' or 'playlist'

    Returns:
        dict: Album information and track list in a format matching the app structure
    """
    try:
        if content_type == 'album':
            content = ytmusic.get_album(content_id)

            total_duration = sum(
                sum(int(t) * (60 ** i) for i, t in enumerate(reversed(track.get('duration', '0:00').split(':'))))
                for track in content.get('tracks', [])
            )

            album_info = {
                "title": content.get('title', 'Unknown Album'),
                "artist": content.get('artists', [{'name': 'Unknown Artist'}])[0]['name'],
                "coverUrl": content.get('thumbnails', [])[-1]['url'] if content.get('thumbnails') else '',
                "description": content.get('description', ''),
                "duration": total_duration
            }

            tracks = [{
                "id": track.get('videoId', ''),
                "number": index + 1,
                "title": track.get('title', 'Unknown Title'),
                "duration": sum(int(t) * (60 ** i) for i, t in enumerate(reversed(track.get('duration', '0:00').split(':')))),
                "quality": '256Kbps',
                "playUrl": f"https://music.youtube.com/watch?v={track.get('videoId')}" if track.get('videoId') else None
            } for index, track in enumerate(content.get('tracks', []))]

            return {
                "album": album_info,
                "tracks": tracks
            }

        elif content_type == 'playlist':
            content = ytmusic.get_playlist(content_id)

            total_duration = sum(
                sum(int(t) * (60 ** i) for i, t in enumerate(reversed(track.get('duration', '0:00').split(':'))))
                for track in content.get('tracks', [])
            )

            playlist_info = {
                "title": content.get('title', 'Unknown Playlist'),
                "artist": content.get('author', {}).get('name', 'Unknown Creator'),
                "releaseDate": '',
                "coverUrl": content.get('thumbnails', [])[-1]['url'] if content.get('thumbnails') else '',
                "description": content.get('description', ''),
                "duration": total_duration
            }

            tracks = [{
                "id": track.get('videoId', ''),
                "number": index + 1,
                "title": track.get('title', 'Unknown Title'),
                "duration": sum(int(t) * (60 ** i) for i, t in enumerate(reversed(track.get('duration', '0:00').split(':')))),
                "quality": 'HIGH',
                "playUrl": f"https://music.youtube.com/watch?v={track.get('videoId')}" if track.get('videoId') else None
            } for index, track in enumerate(content.get('tracks', []))]

            return {
                "album": playlist_info,
                "tracks": tracks
            }

        else:
            raise ValueError("Invalid content type. Must be 'album' or 'playlist'")
    except Exception as e:
        raise Exception(f"Error fetching {content_type} tracks: {str(e)}")

def search_youtube_music(query, search_type="songs", raw_response=False):
    """
    Search YouTube Music for tracks, albums, playlists, podcasts, or artists.

    Parameters:
        query (str): The search query.
        search_type (str): The type of content to search for.
        raw_response (bool): If True, returns the raw API response for podcasts and episodes.

    Returns:
        list: A list of search results based on the search_type.
    """
    search_type_map = {
        'song': 'songs',
        'album': 'albums',
        'playlist': 'playlists',
        'artist': 'artists',
        'podcast': 'podcasts',
        'episode': 'episodes'
    }

    if search_type not in search_type_map:
        raise ValueError(f"Invalid search type: {search_type}. Valid types are {list(search_type_map.keys())}.")

    search_filter = search_type_map[search_type]
    search_results = ytmusic.search(query, filter=search_filter)

    # Return raw response for podcasts and episodes if requested
    if raw_response and search_type in ['podcast', 'episode']:
        return search_results

    formatted_results = []

    if search_type == 'song':
        for result in search_results:
            if 'album' in result:
                formatted_results.append({
                    "TrackTitle": result.get('title', 'Unknown Title'),
                    "AlbumTitle": result['album'].get('name', 'Unknown Album'),
                    "AlbumCover": result['thumbnails'][-1]['url'],
                    "ArtistName": result['artists'][0].get('name', 'Unknown Artist') if result.get('artists') else 'Unknown Artist',
                    "TrackURL": f"https://music.youtube.com/watch?v={result['videoId']}" if 'videoId' in result else 'N/A',
                    "Explicit": result.get('isExplicit', False)
                })
    elif search_type == 'album':
        for result in search_results:
            formatted_results.append({
                "AlbumTitle": result.get('title', 'Unknown Album'),
                "AlbumCover": result['thumbnails'][-1]['url'] if result.get('thumbnails') else 'N/A',
                "ArtistName": result['artists'][0].get('name', 'Unknown Artist') if result.get('artists') else 'Unknown Artist',
                "AlbumURL": f"https://music.youtube.com/browse/{result['browseId']}" if 'browseId' in result else 'N/A',
                "Explicit": result.get('isExplicit', False),
                "BrowseId": result.get('browseId', 'N/A')
            })
    elif search_type == 'playlist':
        for result in search_results:
            formatted_results.append({
                "PlaylistTitle": result.get('title', 'Unknown Playlist'),
                "PlaylistCover": result['thumbnails'][-1]['url'] if result.get('thumbnails') else 'N/A',
                "Author": result.get('author', 'Unknown Author'),
                "PlaylistURL": f"https://music.youtube.com/browse/{result['browseId']}" if 'browseId' in result else 'N/A',
                "BrowseId": result.get('browseId', 'N/A')
            })
    elif search_type == 'artist':
        for result in search_results:
            formatted_results.append({
                "ArtistName": result.get('artist', result.get('title', 'Unknown Artist')),
                "ArtistCover": result['thumbnails'][-1]['url'] if result.get('thumbnails') else 'N/A',
                "ArtistURL": f"https://music.youtube.com/browse/{result['browseId']}" if 'browseId' in result else 'N/A'
            })
    elif search_type == 'podcast':
        for result in search_results:
            formatted_results.append({
                "PodcastTitle": result.get('title', 'Unknown Podcast'),
                "PodcastCover": result['thumbnails'][-1]['url'] if result.get('thumbnails') else 'N/A',
                "PodcastURL": f"https://music.youtube.com/browse/{result['browseId']}" if 'browseId' in result else 'N/A'
            })
    elif search_type == 'episode':
        for result in search_results:
            formatted_results.append({
                "EpisodeTitle": result.get('title', 'Unknown Title'),
                "EpisodeCover": result['thumbnails'][-1]['url'] if result.get('thumbnails') else 'N/A',
                "Podcast": result['podcast']['name'],
                "EpisodeURL": f"https://music.youtube.com/watch?v={result['videoId']}" if 'videoId' in result else 'N/A',
            })
    return formatted_results

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Search YouTube Music for songs, albums, playlists, podcasts, or artists.")
    parser.add_argument('-q', '--query', help="The search query")
    parser.add_argument('-t', '--type', choices=['song', 'album', 'playlist', 'artist', 'podcast', 'episode'],
                        default='song', help="The type of content to search for")
    parser.add_argument('-r', '--raw', action='store_true',
                        help="Return raw API response for podcasts and episodes")
    parser.add_argument('--get-track-list', help="Get track list for album/ID or playlist/ID (format: 'album/ID' or 'playlist/ID')")

    args = parser.parse_args()

    try:
        if args.get_track_list:
            content_type, content_id = args.get_track_list.split('/')
            if content_type not in ['album', 'playlist']:
                raise ValueError("Track list can only be retrieved for albums or playlists")
            results = get_track_list(content_id, content_type)
        elif args.query:
            results = search_youtube_music(args.query, args.type, args.raw)
        else:
            parser.error("Either --query or --get-track-list must be specified")

        if results:
            print(json.dumps(results, indent=4))
        else:
            print("No results found.")
    except Exception as e:
        print(f"Error: {str(e)}")

File: src/funcs/apis/ytmusicsearchapi.js
const YTMusic = require('ytmusic-api');

const ytmusic = new YTMusic();
let _initialized = false;

ytmusic.client.interceptors.request.use((req) => {
    if (req.url) {
        const [path, qs] = req.url.split('?');
        if (qs) {
            const params = new URLSearchParams(qs);
            const key = params.get('key');
            if (!key || key === 'undefined') {
                params.delete('key');
            }
            const cleaned = params.toString();
            req.url = cleaned ? `${path}?${cleaned}` : path;
        }
    }
    req.headers = req.headers || {};
    if (!req.headers['Origin'])  req.headers['Origin']  = 'https://music.youtube.com';
    if (!req.headers['Referer']) req.headers['Referer'] = 'https://music.youtube.com/';
    return req;
});

async function ensureInitialized() {
    if (_initialized) return;

    try {
        await ytmusic.initialize();
        if (ytmusic.config?.INNERTUBE_CLIENT_VERSION) {
            _initialized = true;
            ytmusic.initialize = async () => {};
            return;
        }
    } catch (_) {}

    const fallback = {
        INNERTUBE_API_VERSION: 'v1',
        INNERTUBE_CLIENT_NAME: 'WEB_REMIX',
        INNERTUBE_CLIENT_VERSION: '1.20240101.01.00',
        INNERTUBE_CONTEXT_CLIENT_NAME: 67,
        GL: 'US',
        HL: 'en',
    };

    try {
        const res = await fetch('https://music.youtube.com/', {
            headers: {
                'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/121.0.0.0 Safari/537.36',
                'Accept-Language': 'en-US,en;q=0.9',
            },
        });
        const html = await res.text();
        const pick = (key) => {
            const m = html.match(new RegExp('"' + key + '"\\s*:\\s*"([^"]+)"'));
            return m ? m[1] : null;
        };
        ytmusic.config = {
            ...ytmusic.config,
            ...fallback,
            INNERTUBE_API_KEY:      pick('INNERTUBE_API_KEY')      || '',
            INNERTUBE_API_VERSION:  pick('INNERTUBE_API_VERSION')  || fallback.INNERTUBE_API_VERSION,
            INNERTUBE_CLIENT_NAME:  pick('INNERTUBE_CLIENT_NAME')  || fallback.INNERTUBE_CLIENT_NAME,
            INNERTUBE_CLIENT_VERSION: pick('INNERTUBE_CLIENT_VERSION') || fallback.INNERTUBE_CLIENT_VERSION,
            GL: pick('GL') || fallback.GL,
            HL: pick('HL') || fallback.HL,
        };
    } catch (_) {
        ytmusic.config = { ...ytmusic.config, ...fallback, INNERTUBE_API_KEY: '' };
    }

    _initialized = true;
    ytmusic.initialize = async () => {};
}

function parseDuration(durationStr) {
    if (!durationStr) return 0;
    const parts = durationStr.split(':').reverse();
    let seconds = 0;
    for (let i = 0; i < parts.length; i++) {
        seconds += parseInt(parts[i]) * Math.pow(60, i);
    }
    return seconds;
}

async function getTrackList(contentId, contentType) {
    await ensureInitialized();
    try {
        let content;
        if (contentType === 'album') {
            content = await ytmusic.getAlbum(contentId);
            const totalDuration = content.songs.reduce((sum, track) => sum + (track.duration || 0), 0);
            const albumInfo = {
                title: content.name || 'Unknown Album',
                artist: content.artist?.name || 'Unknown Artist',
                coverUrl: content.thumbnails?.[content.thumbnails.length - 1]?.url || '',
                description: content.description || '',
                duration: totalDuration
            };
            const tracks = content.songs.map((track, index) => ({
                id: track.videoId || '',
                number: index + 1,
                title: track.name || 'Unknown Title',
                duration: track.duration || 0,
                quality: '256Kbps',
                playUrl: track.videoId ? `https://music.youtube.com/watch?v=${track.videoId}` : null
            }));
            return { album: albumInfo, tracks };
        } else if (contentType === 'playlist') {
            const browseId = contentId.startsWith('VL') ? contentId : 'VL' + contentId;
            const rawData = await ytmusic.constructRequest('browse', { browseId });

            const headerItems = [];
            _findAll(rawData, 'musicResponsiveHeaderRenderer', headerItems, 0);
            const header = headerItems[0] || {};
            const headerTitle = header.title?.runs?.[0]?.text || 'Unknown Playlist';
            const subtitleRuns = header.subtitle?.runs || [];
            const headerYear = subtitleRuns.map(r => r.text).find(t => /^\d{4}$/.test(t?.trim())) || '';
            const headerThumbUrl = (() => {
                const thumbs = header.thumbnail?.musicThumbnailRenderer?.thumbnail?.thumbnails;
                return thumbs ? thumbs[thumbs.length - 1]?.url : '';
            })();

            const renderers = [];
            _findAll(rawData, 'musicResponsiveListItemRenderer', renderers, 0);

            const tracks = renderers.map((item, index) => {
                const videoId = item.playlistItemData?.videoId || null;
                if (!videoId) return null;

                const col0runs = item.flexColumns?.[0]?.musicResponsiveListItemFlexColumnRenderer?.text?.runs || [];
                const col1runs = item.flexColumns?.[1]?.musicResponsiveListItemFlexColumnRenderer?.text?.runs || [];
                const fixedRuns = item.fixedColumns?.[0]?.musicResponsiveListItemFixedColumnRenderer?.text?.runs || [];

                const title   = col0runs.map(r => r.text).join('') || 'Unknown Title';
                const artist  = col1runs.map(r => r.text).join('').trim() || '';
                const durStr  = fixedRuns.map(r => r.text).join('').trim() || '';

                return {
                    id: videoId,
                    number: index + 1,
                    title,
                    artist,
                    duration: parseDuration(durStr),
                    quality: 'HIGH',
                    playUrl: `https://music.youtube.com/watch?v=${videoId}`
                };
            }).filter(Boolean);

            const totalDuration = tracks.reduce((sum, t) => sum + t.duration, 0);
            const playlistInfo = {
                title: headerTitle,
                artist: 'YouTube Music',
                releaseDate: headerYear,
                coverUrl: headerThumbUrl || '',
                description: '',
                duration: totalDuration
            };
            return { album: playlistInfo, tracks };
        } else if (contentType === 'podcast') {
            const rawData = await ytmusic.constructRequest('browse', { browseId: contentId });

            const mf = rawData.microformat?.microformatDataRenderer;
            const thumbs = mf?.thumbnail?.thumbnails;
            const podcastInfo = {
                title: mf?.title || 'Unknown Podcast',
                artist: 'Podcast',
                coverUrl: thumbs ? thumbs[thumbs.length - 1]?.url : '',
                description: mf?.description || '',
                duration: 0
            };

            // Try multiple renderer types — YouTube Music uses different renderers over time
            let episodeItems = [];
            _findAll(rawData, 'musicMultiRowListItemRenderer', episodeItems, 0);

            if (episodeItems.length === 0) {
                _findAll(rawData, 'musicResponsiveListItemRenderer', episodeItems, 0);
            }
            if (episodeItems.length === 0) {
                _findAll(rawData, 'musicTwoRowItemRenderer', episodeItems, 0);
            }

            const tracks = episodeItems.map((item, index) => {
                // Handle musicMultiRowListItemRenderer
                const title = item.title?.runs?.[0]?.text
                    // Handle musicResponsiveListItemRenderer
                    || item.flexColumns?.[0]?.musicResponsiveListItemFlexColumnRenderer?.text?.runs?.[0]?.text
                    // Handle musicTwoRowItemRenderer
                    || item.title?.runs?.[0]?.text
                    || 'Unknown Episode';

                const episodeThumbs = item.thumbnail?.musicThumbnailRenderer?.thumbnail?.thumbnails;
                const thumbUrl = episodeThumbs ? episodeThumbs[episodeThumbs.length - 1]?.url : null;

                const videoIds = [];
                _collectIds(item, { browseIds: [], videoIds }, 0);
                const videoId = videoIds[0] || null;

                return {
                    id: videoId || '',
                    number: index + 1,
                    title,
                    duration: 0,
                    quality: 'HIGH',
                    thumbnail: thumbUrl,
                    playUrl: videoId ? `https://music.youtube.com/watch?v=${videoId}` : null
                };
            }).filter(ep => ep.playUrl);

            return { album: podcastInfo, tracks };
        } else {
            throw new Error("Invalid content type. Must be 'album', 'playlist', or 'podcast'");
        }
    } catch (e) {
        throw new Error(`Error fetching ${contentType} tracks: ${e.message}`);
    }
}

const SEARCH_PARAMS = {
    song:     decodeURIComponent('EgWKAQIIAWoKEAkQAxAEEAoQBQ%3D%3D'),
    album:    decodeURIComponent('EgWKAQIYAWoKEAkQAxAEEAoQBQ%3D%3D'),
    playlist: decodeURIComponent('EgeKAQQoAEABagoQCRADEAQQChAF'),
    artist:   decodeURIComponent('EgWKAQIgAWoKEAkQAxAEEAoQBQ%3D%3D'),
    podcast:  decodeURIComponent('Eg%2BKAQwIABAAGAAgACgAMAFqChAEEAMQCRAFEAo%3D'),
};

function _findAll(obj, key, results, depth) {
    if (depth > 12 || !obj || typeof obj !== 'object') return;
    if (obj[key]) results.push(obj[key]);
    for (const v of Object.values(obj)) _findAll(v, key, results, depth + 1);
}

function _collectIds(obj, ids, depth) {
    if (depth > 12 || !obj || typeof obj !== 'object') return;
    if (obj.browseId) ids.browseIds.push(obj.browseId);
    if (obj.videoId)  ids.videoIds.push(obj.videoId);
    for (const v of Object.values(obj)) _collectIds(v, ids, depth + 1);
}

function _extractTexts(obj, texts, depth) {
    if (depth > 12 || !obj || typeof obj !== 'object') return;
    if (Array.isArray(obj.runs)) obj.runs.forEach(r => r.text && texts.push(r.text));
    for (const v of Object.values(obj)) _extractTexts(v, texts, depth + 1);
}

function _findThumbnails(obj, depth) {
    if (depth > 12 || !obj || typeof obj !== 'object') return null;
    if (obj.thumbnails && Array.isArray(obj.thumbnails)) return obj.thumbnails;
    for (const v of Object.values(obj)) {
        const found = _findThumbnails(v, depth + 1);
        if (found) return found;
    }
    return null;
}

function _parsePodcastItems(rawData) {
    const renderers = [];
    _findAll(rawData, 'musicResponsiveListItemRenderer', renderers, 0);

    return renderers.map(item => {
        const texts = [];
        _extractTexts(item, texts, 0);
        const typeLabel = texts[1] || '';
        if (typeLabel !== 'Podcast') return null;

        const ids = { browseIds: [], videoIds: [] };
        _collectIds(item, ids, 0);

        const podcastBrowseId = ids.browseIds.find(id => id.startsWith('MPSP'))
            || ids.browseIds[0]
            || null;

        const thumbs = _findThumbnails(item, 0);
        const thumbnail = thumbs ? thumbs[thumbs.length - 1]?.url : undefined;

        if (!podcastBrowseId) return null;
        return {
            id: podcastBrowseId,
            title: texts[0] || 'Unknown Podcast',
            thumbnail,
            url: `https://music.youtube.com/browse/${podcastBrowseId}`,
        };
    }).filter(Boolean);
}

async function searchYouTubeMusic(query, searchType = 'song') {
    await ensureInitialized();

    const validTypes = ['song', 'album', 'playlist', 'artist', 'podcast'];
    if (!validTypes.includes(searchType)) {
        throw new Error(`Invalid search type: ${searchType}. Valid types are ${validTypes.join(', ')}.`);
    }

    if (searchType === 'podcast') {
        const rawData = await ytmusic.constructRequest('search', {
            query,
            params: SEARCH_PARAMS.podcast,
        });
        return _parsePodcastItems(rawData);
    }

    if (searchType === 'song') {
        const results = await ytmusic.searchSongs(query);
        return results.map(r => ({
            id: r.videoId,
            title: r.name,
            artist: r.artist?.name,
            album: r.album?.name,
            thumbnail: r.thumbnails?.[r.thumbnails.length - 1]?.url,
            url: r.videoId ? `https://music.youtube.com/watch?v=${r.videoId}` : undefined,
            explicit: false,
            duration: typeof r.duration === 'number' ? r.duration : parseDuration(r.duration),
        })).filter(r => r.id);
    }

    if (searchType === 'artist') {
        const results = await ytmusic.searchArtists(query);
        return results.map(r => ({
            id: r.artistId,
            name: r.name,
            title: r.name,
            thumbnail: r.thumbnails?.[r.thumbnails.length - 1]?.url,
            browseId: r.artistId,
            url: r.artistId ? `https://music.youtube.com/browse/${r.artistId}` : undefined,
        })).filter(r => r.id);
    }

    if (searchType === 'album') {
        const results = await ytmusic.searchAlbums(query);
        return results.map(r => ({
            id: r.albumId,
            title: r.name,
            artist: r.artist?.name,
            thumbnail: r.thumbnails?.[r.thumbnails.length - 1]?.url,
            browseId: r.albumId,
            url: r.albumId ? `https://music.youtube.com/browse/${r.albumId}` : undefined,
            explicit: false,
        })).filter(r => r.id);
    }

    if (searchType === 'playlist') {
        const results = await ytmusic.searchPlaylists(query);
        return results.map(r => ({
            id: r.playlistId,
            title: r.name,
            owner: r.artist?.name,
            thumbnail: r.thumbnails?.[r.thumbnails.length - 1]?.url,
            browseId: r.playlistId ? `VL${r.playlistId}` : undefined,
            url: r.playlistId ? `https://music.youtube.com/browse/VL${r.playlistId}` : undefined,
        })).filter(r => r.id);
    }

    return [];
}


async function getArtistAlbums(browseId) {
    await ensureInitialized();
    try {
        const items = await ytmusic.getArtistAlbums(browseId);
        return Array.isArray(items) ? items : [];
    } catch (_) {
        return [];
    }
}

module.exports = { getTrackList, searchYouTubeMusic, getArtistAlbums };
